---
title: "Bayes Factor"
author: "RichÃ¨l J.C. Bilderbeek"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Bayes Factor}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Goal

Measure the Bayes factor between JC69 and GTR.

```{r}
library(mcbette)
```

Set a random seed:

```{r}
set.seed(314)
```

BEAST2 must be installed:

```{r}
if (!beastier::is_beast2_installed()) {
  message("BEAST2 must be installed. Tip: use 'beastier::install_beast2()'")
}
```

The BEAST2 package `NS` must be installed:

```{r}
if (!mauricer::is_beast2_ns_pkg_installed()) {
  message(
    "The BEAST2 'NS' package must be installed. ",
    "Tip: use 'mauricer::install_beast2_pkg(\"NS\")'"
  )
}
```

Create data frame:

```{r}
df <- data.frame(
  site_model_name = c("JC69", "GTR"),
  marg_log_lik = c(-1234.567, -23456.78),
  weight = c(0.9, 0.1)
)
```

## Overview

 * Create a data set that *is* JC69
 * Calculate the mar
 * Compare these posteriors using the Bayes factor
 * Interpret the Bayes factor

## Methods

Create an articifical PBD phylogeny with a long time to speciate. That is,
a tree with all speciation event observed early in time (close to the root):
 
```{r fig.width=7}
phylogeny <- ape::read.tree(text = "((((A:12, B:12):1,C:13):1,D:14):1, E:15);")
crown_age <- beautier::get_crown_age(phylogeny)
ape::plot.phylo(phylogeny)
```

Create JC69 alignment and save it to a FASTA file:

```{r fig.width=7}
sequence_length <- 1000

alignment_phydat <- phangorn::simSeq(
  phylogeny,
  l = sequence_length,
  rate = 1.0 / crown_age,
  rootseq = rep(c('a', 'c', 'g', 't'), each = sequence_length / 4)
)
testit::assert(class(alignment_phydat) == "phyDat")
alignment_dnabin <- ape::as.DNAbin(alignment_phydat)

fasta_filename <- tempfile(
  pattern = "bayes_factor_", 
  tmpdir = rappdirs::user_cache_dir(),
  fileext = ".fasta"
)
phangorn::write.phyDat(
  alignment_dnabin,
  file = fasta_filename,
  format = "fasta"
)
image(alignment_dnabin)
```

From that alignment, get the marginal likelihood of the two desired models:


```{r}
if (mauricer::is_beast2_ns_pkg_installed()) {
  inference_model_1 <- create_test_ns_inference_model(
    site_model = beautier::create_jc69_site_model()   
  )
  inference_model_2 <- create_test_ns_inference_model(
    site_model = beautier::create_gtr_site_model()   
  )
  inference_models <- c(list(inference_model_1), list(inference_model_2))
  
  df <- mcbette::est_marg_liks(
    fasta_filename = fasta_filename, 
    inference_models = inference_models
  )
  knitr::kable(df)
}
```

The Bayes factor (usually denoted as K) can be used for model comparison. 
It can be calculated from the marginal likelihoods ratio:

```{r}
if (mauricer::is_beast2_ns_pkg_installed()) {
  marg_lik_jc <- exp(Rmpfr::mpfr(df$marg_log_lik[1], 512))
  marg_lik_gtr <- exp(Rmpfr::mpfr(df$marg_log_lik[2], 512))
  bayes_factor <- marg_lik_jc / marg_lik_gtr
  interpretation <- interpret_bayes_factor(as.numeric(bayes_factor))
  cat(paste(interpretation, "in favor of JC69"))
}
```
